tests/framework/umbra_client_faulty_redzone.c

```diff
@@ -93,73 +97,110 @@ dr_client_main(client_id_t id, int argc, const char *argv[])
-        drreg_reserve_register(drcontext, ilist, where, IF_X86_ELSE(&allowed, NULL),
-                               &scratch) != DRREG_SUCCESS ||
-        drreg_reserve_register(drcontext, ilist, where, NULL, &regaddr) !=
-            DRREG_SUCCESS) {
-        DR_ASSERT(false); /* Can't recover. */
-        return;
-    }
-#ifdef X86
-    drvector_delete(&allowed);
-#endif
-
-    ok = drutil_insert_get_mem_addr(drcontext, ilist, where, ref, regaddr, scratch);
+    ok = drutil_insert_get_mem_addr(drcontext, ilist, where, ref, scratch_reg, scratch_reg2);
     DR_ASSERT(ok);
     /* Save the app address to a well-known spill slot, so that the fault handler
      * can recover if no shadow memory was installed yet.
      */
-    dr_save_reg(drcontext, ilist, where, regaddr, SPILL_SLOT_2);
+    dr_save_reg(drcontext, ilist, where, scratch_reg, SPILL_SLOT_2);
 
-    if (umbra_insert_app_to_shadow(drcontext, umbra_map, ilist, where, regaddr, &scratch,
+    if (umbra_insert_app_to_shadow(drcontext, umbra_map, ilist, where, scratch_reg, &scratch_reg2,
                                    1) != DRMF_SUCCESS) {
         DR_ASSERT(false);
     }
 
     /* Use a displacement of a page size to access ahead and try to hit a faulty
-     * redzone.
+     * redzone. Naturally, this does not guarantee that the red-zone will always
+     * be accessed, but we hope that it will hit at least once to trigger a fault.
      */
     instrlist_meta_preinsert(
         ilist, where,
         INSTR_XL8(XINST_CREATE_store_1byte(
                       drcontext,
-                      OPND_CREATE_MEM8(regaddr, dr_page_size() /* enter redzone */),
-                      opnd_create_reg(reg_resize_to_opsz(scratch, OPSZ_1))),
+                      OPND_CREATE_MEM8(scratch_reg, dr_page_size() /* try enter redzone */),
+                      opnd_create_reg(reg_resize_to_opsz(scratch_reg2, OPSZ_1))),
                   instr_get_app_pc(where)));
+}
 
-    if (drreg_unreserve_register(drcontext, ilist, where, regaddr) != DRREG_SUCCESS ||
-        drreg_unreserve_register(drcontext, ilist, where, scratch) != DRREG_SUCCESS ||
-        drreg_unreserve_aflags(drcontext, ilist, where) != DRREG_SUCCESS)
-        DR_ASSERT(false);
+static bool
+instr_uses_mem(instr_t *where)
+{
+    int i;
+    for (i = 0; i < instr_num_srcs(where); i++) {
+        if (opnd_is_memory_reference(instr_get_src(where, i)))
+          return true;
+    }
+    for (i = 0; i < instr_num_dsts(where); i++) {
+        if (opnd_is_memory_reference(instr_get_dst(where, i)))
+            return true;
+    }
+
+    return false;
 }
 
 static dr_emit_flags_t
 event_app_instruction(void *drcontext, void *tag, instrlist_t *ilist, instr_t *where,
                       bool for_trace, bool translating, void *user_data)
 {
+    reg_id_t scratch_reg;
+    reg_id_t scratch_reg2;
+    instr_t *label = NULL;
+    bool uses_mem = instr_uses_mem(where);
+
+    if (uses_mem){
+
+    #ifdef X86
```

Comment:
We always have # in the first column.  Ditto below.

Reply:
Done.